### PR TITLE
Separate the Connection establishment from the Connection functionality

### PR DESCRIPTION
**Pull request by jennifer-at-improbable-io**
**Contributions**: We are not currently taking public contributions - see our [contributions](CONTRIBUTING.md) policy. However, we are accepting issues and we do want your [feedback](../README.md#give-us-feedback).

-------

#### Description
This should have no functional change, it is just reoganization.

The existing SpatialWorkerConnection used to both be responsible for negotiating the connection with SpatialOS and also responsible for communicating data down that connection. This change separates the functionality into the SpatialConnectionManager (which creates the connection) and the SpatialWorkerConnection (which now just pipes data across the connection).

#### Release note
REQUIRED: Add a release note to the `##Unreleased` section of CHANGELOG.md. You can find guidance for writing useful release notes [here](../SpatialGDK/Extras/internal-documentation/how-to-write-good-release-notes.md). Documentation changes are exempt from this requirement.

#### Tests
How did you test these changes prior to submitting this pull request?
What automated tests are included in this PR?

STRONGLY SUGGESTED: How can this be verified by QA?

#### Documentation
How is this documented (for example: release note, upgrade guide, feature page, in-code documentation)?

#### Primary reviewers
If your change will take a long time to review, you can name at most two primary reviewers who are ultimately responsible for reviewing this request. @ mention them.

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialLoadBalanceEnforcer.cpp [hunk 1]
```diff
@@ -21,7 +21,7 @@ SpatialLoadBalanceEnforcer::SpatialLoadBalanceEnforcer(const PhysicalWorkerName&
 
 void SpatialLoadBalanceEnforcer::OnAuthorityIntentComponentUpdated(const Worker_ComponentUpdateOp& Op)
 {
-	check(Op.update.component_id == SpatialConstants::AUTHORITY_INTENT_COMPONENT_ID)
```
Comment (jennifer-at-improbable-io) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialLoadBalanceEnforcer.cpp]: The changes to this file are just some debugging I added when helping Jacques diagnose a problem, but I thought they were useful so I left them in.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialConnectionManager.cpp [hunk 2]
```diff
@@ -0,0 +1,458 @@
...
+		Worker_Connection* NewCAPIWorkerConnection = Worker_ConnectionFuture_Get(ConnectionFuture, nullptr);
+		Worker_ConnectionFuture_Destroy(ConnectionFuture);
+
+		AsyncTask(ENamedThreads::GameThread, [WeakSpatialConnectionManager, NewCAPIWorkerConnection]
+		{
+			USpatialConnectionManager* SpatialConnectionManager = WeakSpatialConnectionManager.Get();
+
+			if (Worker_Connection_IsConnected(NewCAPIWorkerConnection))
+			{
+				SpatialConnectionManager->WorkerConnection = NewObject<USpatialWorkerConnection>();
```
Comment (jennifer-at-improbable-io) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialConnectionManager.cpp]: The next 10 lines are the only meaningful change from the original connection code in SpatialWorkerConnection.cpp.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialConnectionManager.cpp [hunk 3]
```diff
@@ -0,0 +1,458 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
```
Comment (jennifer-at-improbable-io) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialConnectionManager.cpp]: This file is almost exclusively moved from SpatialWorkerConnection.cpp except for name changes and where tagged below.

File: SpatialGDK/Source/SpatialGDK/Public/Interop/Connection/SpatialConnectionManager.h [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+enum class ESpatialConnectionType
+{
+	Receptionist,
+	LegacyLocator,
+	Locator,
+	DevAuthFlow
+};
+
+UCLASS()
+class SPATIALGDK_API USpatialConnectionManager : public UObject
```
Comment (samiwh) [inline SpatialGDK/Source/SpatialGDK/Public/Interop/Connection/SpatialConnectionManager.h]: Possibly out of scope for this PR, as this is a reorganisation of existing code rather than a rewrite. If so feel free to resolve and the discussion can happen offline.

As the connection manager is meant to create the connection, does it need to be a UObject and is it right that it should own the connection?

It looks like this only reason this is a UObject is because the worker connection is a UObject which I'm not sure needs to be either.

Could it instead be some object we build up some function that returns what is effectively a future. We check the future each tick then eventually give the resulting connection to the workerconnection object, or retry if it fails (which I think we don't currently have any ability to do anyway.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 5]
```diff
@@ -1747,7 +1748,7 @@ void USpatialNetDriver::TickFlush(float DeltaTime)
 
 	if (SpatialGDKSettings->bRunSpatialWorkerConnectionOnGameThread)
 	{
-		if (Connection != nullptr && Connection->IsConnected())
+		if (Connection != nullptr)
```
Comment (samiwh) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: My only concern with doing this everywhere is that it's less explicit. And generally when seeing !=nullptr I'm not entirely sure why something might be nullptr. It could be that using ConnectionManager->IsConnected() would be better but then you want to be querying the workerconnection unless you are the game instance.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialLoadBalanceEnforcer.cpp [hunk 6]
```diff
@@ -42,7 +42,7 @@ void SpatialLoadBalanceEnforcer::AuthorityChanged(const Worker_AuthorityChangeOp
 			// TODO(zoning): There are still some entities being created without an authority intent component.
 			// For example, the Unreal created worker entities don't have one. Even though those won't be able to
 			// transition, we should have the intent component on them for completeness.
-		 	UE_LOG(LogSpatialLoadBalanceEnforcer, Log, TEXT("Requested authority change for entity without AuthorityIntent component. EntityId: %lld"), AuthOp.entity_id);
+		 	UE_LOG(LogSpatialLoadBalanceEnforcer, Log, TEXT("(%s) Requested authority change for entity without AuthorityIntent component. EntityId: %lld"), *WorkerId, AuthOp.entity_id);
```
Comment (samiwh) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialLoadBalanceEnforcer.cpp]: Nit - I think we prefer to put dynamic parts of the message at the end rather than at the beginning. We are very inconsistent about this but try to do it for new stuff.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialVirtualWorkerTranslator.cpp [hunk 7]
```diff
@@ -22,7 +22,7 @@ const PhysicalWorkerName* SpatialVirtualWorkerTranslator::GetPhysicalWorkerForVi
 
 void SpatialVirtualWorkerTranslator::ApplyVirtualWorkerManagerData(Schema_Object* ComponentObject)
 {
-    UE_LOG(LogSpatialVirtualWorkerTranslator, Log, TEXT("ApplyVirtualWorkerManagerData"));
+    UE_LOG(LogSpatialVirtualWorkerTranslator, Log, TEXT("(%d) ApplyVirtualWorkerManagerData"), LocalVirtualWorkerId);
```
Comment (samiwh) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialVirtualWorkerTranslator.cpp]: Potential driveby (realise this change is just making it saying which worker actually ran this code)
What is this log for? This looks like a specific debugging log that got left in by accident.

Comment (samiwh) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialVirtualWorkerTranslator.cpp]: If this is useful we should probably make the text saying a complete sentence.

File: SpatialGDK/Source/SpatialGDK/Public/Interop/Connection/SpatialConnectionManager.h (hunk 4 above)
Comment (jennifer-at-improbable-io) [reply SpatialGDK/Source/SpatialGDK/Public/Interop/Connection/SpatialConnectionManager.h]: I agree that it doesn't need to be a UObject, but I explicitly didn't want to do that in this PR because I wanted to be able to isolate ownership/memory management issues with the changing objects because of the restructuring, as opposed to as a result of a UObject conversion. I'd be very interested in doing that as a follow-up.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp (hunk 5 above)
Comment (jennifer-at-improbable-io) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: Really except for startup, this should always be valid.   (There's an open reconnect discussion, but I think that's outside of the scope.)  Would you feel better if this method (and one other in the tick processing) just had an early out for no connection and then we assume that it's valid any other time?

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialLoadBalanceEnforcer.cpp (hunk 6 above)
Comment (jennifer-at-improbable-io) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialLoadBalanceEnforcer.cpp]: That's actually the opposite of most of the debugging in the load balancing. I was actually removing all of those, but because Squid doesn't have logging (and so the SpatialLogOutput is empty) I was adding it back in for debugging on Squid.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialVirtualWorkerTranslator.cpp (hunk 7 above)
Comment (jennifer-at-improbable-io) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialVirtualWorkerTranslator.cpp]: It doesn't make sense as a single message, but it's a header for the data below, so you get:

(3) ApplyVirtualWorkerManagerData
Translator assignment: 1 - Unreal35238
Translator assignment: 2 - Unreal23412343
etc

So the message just tells you which worker has received the translation mapping and what they think it is.

Comment (samiwh) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialVirtualWorkerTranslator.cpp]: Ah I see. Thanks.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialLoadBalanceEnforcer.cpp (hunk 6 above)
Comment (samiwh) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialLoadBalanceEnforcer.cpp]: I think there were quite a few messages that only really make sense to put things at the end. Generally when you want a load of metadata like, entity ID, actor name, request ID, and status message all in one log.

It got briefly discussed a while ago as something to adopt although I wouldn't be surprised if not enforced. The rational was being able to copy and find a message without worrying about what might be dynamic (although it's often obvious).
It might be worth deciding on this offline though if we've been inconsistent recently.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp (hunk 5 above)
Comment (samiwh) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: I think leaving the check in is fine for now. It was more on the != nullptr, vs ->IsConnected. But If this is going to be a pain I think this is fine.

File: SpatialGDK/Source/SpatialGDK/Public/Interop/Connection/SpatialConnectionManager.h [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    /// It will be triggered when receiving login tokens using the development authentication flow inside SpatialWorkerConnection.
+    /// @param Callback - callback function.
+	void RegisterOnLoginTokensCallback(const LoginTokenResponseCallback& Callback) {LoginTokenResCallback = Callback;}
+
+	void Connect(bool bConnectAsClient, uint32 PlayInEditorID);
+
+	FORCEINLINE bool IsConnected() { return bIsConnected; }
+
+	PhysicalWorkerName GetWorkerId() const;
+	const TArray<FString>& GetWorkerAttributes() const;
```
Comment (aleximprobable) [inline SpatialGDK/Source/SpatialGDK/Public/Interop/Connection/SpatialConnectionManager.h]: Do we need these functions in `SpatialConnectionManager`?

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 9]
```diff
@@ -237,44 +238,47 @@ void USpatialNetDriver::InitiateConnectionToSpatialOS(const FURL& URL)
...
+	ConnectionManager->OnFailedToConnectCallback.BindUObject(this, &USpatialNetDriver::OnConnectionToSpatialOSFailed);
 
 	// If this is the first connection try using the command line arguments to setup the config objects.
 	// If arguments can not be found we will use the regular flow of loading from the input URL.
 	FString SpatialWorkerType = GetGameInstance()->GetSpatialWorkerType().ToString();
 	if (!GameInstance->GetFirstConnectionToSpatialOSAttempted())
 	{
 		GameInstance->SetFirstConnectionToSpatialOSAttempted();
-		if (!Connection->TrySetupConnectionConfigFromCommandLine(SpatialWorkerType))
+		if (!ConnectionManager->TrySetupConnectionConfigFromCommandLine(SpatialWorkerType))
```
Comment (aleximprobable) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: Can we move this logic to `SpatialConnectionManager`? Or is it something for the future PRs?

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialGameInstance.cpp [hunk 10]
```diff
@@ -73,20 +73,15 @@ bool USpatialGameInstance::HasSpatialNetDriver() const
 
 void USpatialGameInstance::CreateNewSpatialWorkerConnection()
```
Comment (aleximprobable) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialGameInstance.cpp]: Should this be renamed to `CreateNewSpatialWorkerConnectionManager` now?

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialLoadBalanceEnforcer.cpp (hunk 6 above)
Comment (jennifer-at-improbable-io) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialLoadBalanceEnforcer.cpp]: Happy to discuss offline. In this case, the ID is being used as a way to filter multiple messages from different workers, so putting it at the beginning mirrors the SpatialLogOutput.

File: SpatialGDK/Source/SpatialGDK/Public/Interop/Connection/SpatialConnectionManager.h (hunk 8 above)
Comment (jennifer-at-improbable-io) [reply SpatialGDK/Source/SpatialGDK/Public/Interop/Connection/SpatialConnectionManager.h]: I removed those and pointed a couple of places at the connection's implementation.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialGameInstance.cpp (hunk 10 above)
Comment (jennifer-at-improbable-io) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialGameInstance.cpp]: Good point, done.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp (hunk 9 above)
Comment (jennifer-at-improbable-io) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: I feel like there's another effort around reconnect that will probably deal with this code, so I'd rather leave it for then.
